### ISSUE TITLE
Cannot assign same type object to explicitly-typed generic descriptor

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1045]
**Describe the bug**
Attempting to overwrite an explicitly-typed, parametrised non-data descriptor (a descriptor declared without a `__set__` method) on class instances with an object of the same type as the return value of the descriptor's `__get__` produces an error.  Unannotated descriptors where the type can be inferred from `__init__` do not exhibit the same behaviour.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `bar: Column[str] = Column(str)` and `baz = Column(str)` would behave in the same way.

**Screenshots or Code**
```python
# pyright: strict

from typing import Any, Generic, Type, TypeVar, overload


_T = TypeVar('_T')


class Column(Generic[_T]):
    def __init__(self, type: Type[_T]) -> None: ...
    @overload
    def __get__(self, instance: None, owner: Any) -> 'Column[_T]': ...
    @overload
    def __get__(self, instance: object, owner: Any) -> _T: ...
    # def __set__(self, instance: object, value: Any) -> _T: ...


class Foo:
    bar: Column[str] = Column(str)
    baz = Column(str)

Foo.bar  # Column[str]
Foo().bar  # Just str
Foo.baz  # Column[str]
Foo().baz  # Just str

foo = Foo()
foo.bar = ''  # Cannot assign member "bar" for type "Foo"
              #   Expression of type "Literal['']" cannot be assigned to member "bar" of class "Foo"
              #     "Literal['']" is incompatible with "Column[str]"
foo.baz = ''  # Ok


```

**VS Code extension or command-line**
Extensiom 1.1.73

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

This turned out to be a challenging bug to fix. The problem is that assignments to members are used to infer the type of a member when it has no declared type. But type inference for an unannotated member should take into account only the assignments used within the class definition, not outside of it. In your case, `foo.baz = ''` was outside of `Foo`'s class definition, and the inferred type of `baz` should have been based only on the assignment(s) within the `Foo` class definition.

**Comment by layday** [comment]
Thank you, I assume that means that both of these will produce an error in the next version of Pyright?  mypy doesn't report any issues, though I don't know whether that is intentional.

**Comment by erictraut** [comment]
Yes, that's correct — pyright will produce an error in both cases. It's a type violation, after all. I'm guessing that it's just an omission in mypy.

**Comment by layday** [comment]
But `bar` is a string on instances of `Foo`, right?  Why would it be a violation to assign a string to it?  mypy actually tests for this [here](https://github.com/python/mypy/blob/05571b17a29ef83708ffaa02756e32e60621f69f/test-data/unit/check-classes.test#L1349-L1357).

**Comment by erictraut** [comment]
No, `bar` is not a string on instances of `Foo`. Instances of `Foo` don't have a symbol `bar` at all. The _class_ `Foo` has a symbol `bar` pointing to an instance of `Column[str]`.

You can see this by adding the following code to your sample above:
```python
print(Foo.__dict__.get("baz"))
print(foo.__dict__.get("baz"))
foo.baz = "BAZ!"
print(foo.__dict__.get("baz"))
```

The output is:
```
<__main__.Column object at 0x7fbb6006f880>
None
BAZ!
```

So the statement `foo.baz = "BAZ!"` is overwriting the original value of `baz` with an incompatible type. It was originally of type `Column[str]`, and it is being overwritten with `str`. That's a type violation.

If `Column[str]` had a defined `__set__` method, then it would be acceptable because the statement `foo.baz = "BAZ!"` would not be overwriting `baz`. Instead, it would be calling the setter for the `Column[str]` instance that `baz` refers to.

**Comment by layday** [comment]
I find descriptors rather daunting and perhaps I am misunderstanding, but `foo.baz = "BAZ!"` shadows `baz` from the class.  If you were to `del foo.baz`, `foo.baz` would revert to calling `Foo.__dict__['baz'].__get__(foo, Foo)`, which, in this instance, would yield `None` because it's not implemented.  In both cases, `foo.baz` as typed is a string and the `Column` is not exposed on instances.  This is in effect what `def __get__(self, instance: object, owner: Any) -> _T: ...` says: on instances of `owner` (`Foo`), return `_T` (a string).  Is this materially different from

```py
class Foo:
    bar: str
```

for instances of `Foo`?  mypy seems to think that it isn't.

**Comment by erictraut** [comment]
> In both cases, `foo.baz` as typed is a string

No, a `Column` object that implements a `__get__` method and a `str` object are not the same. They can have different behaviors. I don't think it's safe for a type checker to assume that they're the same.

Consider the following:

```python
class StringSwapper:
    def __init__(self, value: str):
        self._value = value

    def __get__(self, instance: object, owner: Any) -> str:
        return self._value[::-1]

class Foo1:
    bar: str = "Hello World"

class Foo2:
    bar: StringSwapper = StringSwapper("Hello World")

f1 = Foo1()
print(f1.bar)
f1.bar = "Hello World"
print(f1.bar)

f2 = Foo2()
print(f2.bar)
f2.bar = "Hello World"
print(f2.bar)
```

**Comment by layday** [comment]
I can see how your approach is safer.  I suppose the overarching issue here is that descriptors are used to express the kind of relationship where a class with a symbol `x` is of type `y` on the class object and `z` on instances, even when they might not be implemented as descriptors at runtine.  In any case, it was [trivial](https://github.com/layday/sqlalchemy-stubs/commit/b4be519502d30c4e89c19e5975031db92c607f68) to modify SQLAlchemy, where I first encountered this issue, to supress errors on assignment.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.74, which I just published. It will also be included in the next release of Pylance.
